src/module.c

```diff
@@ -4314,7 +4306,7 @@ int RM_ListInsert(RedisModuleKey *key, long index, RedisModuleString *value) {
 int RM_ListDelete(RedisModuleKey *key, long index) {
     if (moduleListIteratorSeek(key, index, REDISMODULE_WRITE)) {
         listTypeDelete(key->iter, &key->u.list.entry);
-        moduleDelKeyIfEmpty(key);
+        if (!moduleDelKeyIfEmpty(key)) moduleFreeKeyIterator(key);
```

Comment:
IIRC that was one of the main design efforts of the list iterator module API.
i.e. that someone can iterate on the list and delete some items, without having to start the iteration from scratch each time.
right?

Reply:
@oranagra The iterator is still available after `listTypeDelete`, and it does not have any problems.
The main problem is reusing the entry, when the index of `moduleListIteratorSeek` is equal to the index of last iterator, we will reuse it.
https://github.com/redis/redis/blob/eb6accad40577cd11846c4f5391c6351eb46d7b6/src/module.c#L4118-L4120

@zuiderkwast  I just updated entry the way you said, but it doesn't look like it can be done simply, if we want that much, we also need to update `entry->zi` which will be set to NULL in (quicklistDelEntry() -> quicklistDelIndex()).
I think we should make sure the iterator is available after deleting the element, but not that entry is available.

Reply:
@sundb i must be missing something.
your change calls `moduleFreeKeyIterator` if the key isn't deleted.
so the next call to the list module APIs (e.g. RM_ListGet) will need to re-iterate from the beginning.

Reply:
@oranagra Yes

Reply:
I think we have to do that.
maybe we can also figure out a way to write a test that will fail if iteration are being reset in various cases.

Reply:
@zuiderkwast Sure, let me work on it again tomorow.

Reply:
@zuiderkwast It seems that updating the entry is not feasible.
For example, we now have a list(x, y, z), and then use `index(2)` to delete z. If we update entry it will point to y, but then `key->u.list.index` will still be 2(origin index).
Unless we update `key->u.list.index` at the same time at the end, this must lead to more complexity.

Reply:
Ok... So why can't we update `key->u.list.index` in `RM_ListDelete` if we delete the last element? It is a slightly more complex because we rely on `listTypeDelete` to update the iterator in this way, but checking that we delete the last element of the list is simple...

    if (index == -1 || index == listTypeLength(key->value) - 1) {
        /* it's the last element */
    }

Reply:
Please take a look at my draft commit(https://github.com/redis/redis/pull/11383/commits/40bc981be5ac3ee33a9a9bfb377c915d2888cfe8), It works, but I don't really like this implement.
The main reason it is inconsistent with the deletion elsewhere is actually, as in the following case.
As I said in the comment above, if a list (x, y, z), when we delete z, we should end up with p pointing to NULL, not y
Both quicklistDelIndex and lpDelete were originally handled this way, but now we need to move p to the last element after deletion.

Reply:
disclaimer: i'm sorry if i'm not following the details closely, please fill in my gaps.

i agree it looks very odd that if you delete the last entry you move to the last entry rather than NULL, i think it can easily cause bugs for the unsuspecting caller we well...

can you please explain to me again how this is related to plain nodes? and how it used to work before (if at all)?
or is it that it was always buggy, but only with plain nodes it crashes?

does the problem only happens on an iterator that goes backwards and the very last element in the list us deleted?
if so, as far as the module api cares, we can reset the iterator in this case, since seeking to the last index is quick.
i assume we also need that per-listpack, so maybe a boolean argument to the delete function or the iterator can help?

Reply:
> can you please explain to me again how this is related to plain nodes? and how it used to work before (if at all)? or is it that it was always buggy, but only with plain nodes it crashes?

Until then, deleting a node at the same time as we delete an entry will crash, and has nothing related to do with plain.
But now we need to update the entry, so we have to do the same thing as quicklistNext (decompress and determine if it's plain).
https://github.com/redis/redis/blob/3193f086ca0e167e89b6c5cf14133c03213b8378/src/quicklist.c#L1321-L1325

> 
> does the problem only happens on an iterator that goes backwards and the very last element in the list us deleted? if so, as far as the module api cares, we can reset the iterator in this case, since seeking to the last index is quick. i assume we also need that per-listpack, so maybe a boolean argument to the delete function or the iterator can help?

Yes, the problem now is the question of whether deleting the last element updates the entry.

@zuiderkwast Sorry for my mistake, you may be wondering why I updated the entry to the last entry after deleting the last element.
The main reason is that `index` can be -1, if a list (x,y,z), after deleting index(-1), we also need to update entry at the same time, because `key->u.list.index` is always valid.

@oranagra Adding a bool pram seems to be the most logical and simple way.

Please see commit [`0fa8058` (#11383)](https://github.com/redis/redis/pull/11383/commits/0fa80583d334999c54cb76bdd476cd9164778fa7).